### PR TITLE
SAT Physics

### PR DESCRIPTION
**Pull request by nickelpro**
This is a complete rework of Physics, please test and review

### CODE REVIEW

File: spock/utils.py [hunk 1]
```diff
@@ -41,18 +41,34 @@ def get_dict(self):
         return d
 
 
-class BoundingBox(object):
-    def __init__(self, w, h, d=None, offset=(0, 0, 0)):
-        self.x = offset[0]
-        self.y = offset[1]
-        self.z = offset[2]
```
Comment (Gjum) [inline spock/utils.py]: Do all bounding boxes start at `(0,0,0)` now?

File: spock/plugins/helpers/physics.py [hunk 2]
```diff
@@ -79,87 +77,91 @@ class PhysicsPlugin(PluginBase):
...
+        self.apply_drag()
+        mtv = self.get_mtv()
+        self.pos.on_ground = mtv.y > 0
+        self.apply_vector(mtv)
 
-    def apply_horizontal_drag(self):
+    def apply_drag(self):
+        self.vec.y -= self.vec.y * PLAYER_ENTITY_DRG
         self.vec.x -= self.vec.x * PLAYER_GND_DRG
         self.vec.z -= self.vec.z * PLAYER_GND_DRG
```
Comment (Gjum) [inline spock/plugins/helpers/physics.py]: Is this the same when not `on_ground`?

File: spock/plugins/helpers/physics.py [hunk 3]
```diff
@@ -79,87 +77,91 @@ class PhysicsPlugin(PluginBase):
...
-        p.z = p.z + self.vec.z
+    def apply_vector(self, mtv):
+        self.vec += mtv
+        self.pos += self.vec
+
+    def gen_block_position(self, pos):
+        x = math.floor(pos.x)
+        y = math.floor(pos.y)
+        z = math.floor(pos.z)
+        return Position(x, y, z)
```
Comment (Gjum) [inline spock/plugins/helpers/physics.py]: Use `pos.floor()`

File: spock/plugins/helpers/physics.py [hunk 4]
```diff
@@ -39,6 +39,7 @@
 
 import logging
 import math
+import queue
```
Comment (Gjum) [inline spock/plugins/helpers/physics.py]: Py27: `ImportError: No module named queue`
https://travis-ci.org/SpockBotMC/SpockBot/jobs/76150340#L134

File: spock/utils.py [hunk 5]
```diff
@@ -41,18 +41,34 @@ def get_dict(self):
...
+    def h(self, value):
+        self.vector[1] = value
+
+    @property
+    def d(self):
+        return self.vector[2]
+
+    @d.setter
+    def d(self, value):
+        self.vector[2] = value
```
Comment (Gjum) [inline spock/utils.py]: To avoid code duplication, you could just set `w, h, d` to `x, y, z`

File: spock/plugins/helpers/physics.py [hunk 6]
```diff
@@ -79,87 +77,91 @@ class PhysicsPlugin(PluginBase):
...
     }
+    unit_vectors = Vector3(1, 0, 0), Vector3(0, 1, 0), Vector3(0, 0, 1)
 
     def __init__(self, ploader, settings):
         super(PhysicsPlugin, self).__init__(ploader, settings)
-
         self.vec = Vector3(0.0, 0.0, 0.0)
-        # wiki says 0.6 but I made it 0.8 to give a little wiggle room
-        self.playerbb = BoundingBox(0.8, 1.8)
+        self.playerbb = BoundingBox(0.6, 1.8)
```
Comment (Gjum) [inline spock/plugins/helpers/physics.py]: Can this be more accessible somehow? I want to use it in the InteractPlugin.

File: spock/plugins/helpers/physics.py (hunk 2 above)
Comment (nickelpro) [inline spock/plugins/helpers/physics.py]: No, the acceleration and drag both on and off ground is outright wrong currently. That will need to be fixed but for now this is Close Enough™ to allow for implementation of other plugins.

**Comment by nickelpro** [comment]
I'm signing off on this as Good Enough For Government Work. @gamingrobot please merge if you aren't religiously opposed to anything in this PR

There are lots of edge cases that are broken but it's about 1000% better than what's currently in mainline and sitting in this PR isn't going to make this implementation any better. The next implementation will need to use a projection mesh for testing player movement and will reuse a lot of this code but it's a major refactor and I'm not certain on how to do it yet (Spock needs a math person).

**Pull request merged**